Lib/imaplib.py

```diff
@@ -1279,6 +1389,251 @@ def print_log(self):
+        # argument and block until the next response arrives, because
+        # Windows select() only works on sockets.
+
+        imap = self._imap
+        if imap.state != 'IDLING':
+            raise imap.error('_pop() only works during IDLE')
+
+        if imap._idle_responses:
+            # Response is ready to return to the user
+            resp = imap._idle_responses.pop(0)
+            if __debug__ and imap.debug >= 4:
+                imap._mesg(f'idle _pop({timeout}) de-queued {resp[0]}')
+            return resp
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _pop({timeout})')
+
+        if not self._wait(timeout):
+            if __debug__ and imap.debug >= 4:
+                imap._mesg(f'idle _pop({timeout}) done')
+            return default
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _pop({timeout}) reading')
+        imap._get_response()  # Reads line, calls _append_untagged()
+        resp = imap._idle_responses.pop(0)
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _pop({timeout}) read {resp[0]}')
+        return resp
+
+    def burst(self, interval=0.1):
+        """Yield a burst of responses no more than 'interval' seconds apart
+
+        :param interval:    Time limit for each response after the first
+                            (The IDLE context's maximum duration is
+                            respected when waiting for the first response.)
+        :type interval:     int|float
+
+        This generator retrieves the next response along with any
+        immediately available subsequent responses (e.g. a rapid series of
+        EXPUNGE responses after a bulk delete) so they can be efficiently
+        processed as a batch instead of one at a time.
+
+        Represents responses as (type, datum) tuples, just as when
+        iterating directly on the context manager.
+
+        Example:
+
+        with imap.idle() as idler:
+            batch = list(idler.burst())
+            print(f'processing {len(batch)} responses...')
+
+        Produces no responses and returns immediately if the IDLE
+        context's maximum duration (the 'duration' argument) has elapsed.
+        Callers should plan accordingly if using this method in a loop.
+
+        Note: Windows IMAP4_stream connections will ignore the interval
+        argument, yielding endless responses and blocking indefinitely
+        for each one, because Windows select() only works on sockets.
+        It is therefore advised not to use this method with an IMAP4_stream
+        connection on Windows.
+        """
+        try:
+            yield next(self)
+        except StopIteration:
+            return
+
+        start = time.monotonic()
+
+        yield from iter(functools.partial(self._pop, interval, None), None)
+
+        if self._duration is not None:
+            elapsed = time.monotonic() - start
+            self._duration = max(self._duration - elapsed, 0)
```

Comment:
This looks a bit complicated. You subtract the elapsed time since `start` from the (remaining) duration, but this allows for small amounts of drift to creep in due to time unaccounted for.

One algorithm I've used is to calculate the (absolute) end time (as `time.monotonic() + self._duration`) once, when the context manager is entered, and then only call `time.monotonic() just before calling `select()`, or actually, before calling `_wait()`, to compute the timeout to pass to `select()`.